[SPARK-5644] [Core]Delete tmp dir when sc is stop

Sephiroth-Lin opened the pull request:
When we run driver as a service, and for each time we run job we only call sc.stop, then will not delete tmp dir create by HttpFileServer and SparkEnv, it will be deleted until the service process exit, so we need to delete these tmp dirs when sc is stop directly.

Sephiroth-Lin committed:
Delete tmp dir when sc is stop

When we run driver as a service, and for each time we run job we only call sc.stop, then will not delete tmp dir create by HttpFileServer and SparkEnv, it will be deleted until the service process exit

Sephiroth-Lin committed:
Delete tmp dir when sc is stop

When we run driver as a service, and for each time we run job we only call sc.stop, then will not delete tmp dir create by HttpFileServer and SparkEnv, it will be deleted until the service process exit

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
You'll need a JIRA for this. I think it's probably a harmless change, but, a process contains one SparkContext, and isn't generally doing anything else but Spark, so what problem does this cause in practice?

Sephiroth-Lin commented:
@srowen we run a process as a service which will not stop. In this service process we will create SparkContext and run job and then stop it, because we only call sc.stop but not exit this service process so the tmp dirs created by HttpFileServer and SparkEnv will not be deleted after SparkContext is stopped, and this will lead to creating too many tmp dirs if we create many SparkContext to run job in this service process.

Sephiroth-Lin committed:
update comment

Sephiroth-Lin committed:
update comment

srowen commented on core/src/main/scala/org/apache/spark/HttpFileServer.scala:
```diff
@@ -50,6 +50,15 @@ private[spark] class HttpFileServer(
+    // If we only stop sc, but the driver process still run as a services then we need to delete 
+    // the tmp dir, if not, it will create too many tmp dirs
+    try {
+      Utils.deleteRecursively(baseDir)
```
This seems OK to me. I would make the error log a warning though as it's non-fatal. Consider using an interpolated string here as you do below.

srowen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -93,6 +93,14 @@ class SparkEnv (
+    // If we only stop sc, but the driver process still run as a services then we need to delete 
+    // the tmp dir, if not, it will create too many tmp dirs
+    try {
+      Utils.deleteRecursively(new File(sparkFilesDir))
```
I'm not sure this one is safe. Looking at the code, `sparkFilesDir` will be the current working directory on executors. Is that really safe to delete recursively?

JoshRosen replied:
I agree; this seems unsafe.  It would be a disaster if we accidentally deleted directories that we didn't create, so we can't delete any path that could point to the CWD.  Instead, we might be able to either ensure that the CWD is a subfolder of a spark local directory (so it will be cleaned up as part of our baseDir cleanup) or just change `sparkFilesDir` to not download files to the CWD (e.g. create a temporary directory in both the driver and executors).

Old versions of the `addFile` API contract said that files would be downloaded to the CWD, but we haven't made that promise since Spark 0.7-ish, I think; we only technically guarantee that SparkFIles.get will return the file paths.

Sephiroth-Lin replied:
@srowen @JoshRosen you are right, as we only need to delete the tmp dir create by driver, so check the sparkFilesDir first to confirm it was created by driver. Thank you.

Sephiroth-Lin committed:
Change log level

make the error log a warning though as it's non-fatal.

Sephiroth-Lin committed:
add dir check before delete

sparkFilesDir is point to the current working dir in executor, and we only need to delete the tmp dir create by driver, for safe, we check it first.

Sephiroth-Lin committed:
format code

Sephiroth-Lin committed:
don't check sparkFilesDir, check executorId

srowen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -93,6 +93,19 @@ class SparkEnv (
+    // the tmp dir, if not, it will create too many tmp dirs.
+    // We only need to delete the tmp dir create by driver, because sparkFilesDir is point to the
+    // current working dir in executor which we do not need to delete.
+    if (SparkContext.DRIVER_IDENTIFIER == executorId) {
```
I see that this condition should indeed match the logic which controls whether this directory is a temp dir, but since the consequence of getting this wrong is pretty significant, I wonder if we should be more conservative. For example, save off the `isDriver` flag and use it in both places to be sure that these are exactly in sync? or maintain a separate reference to the temp dir that was created, which is `None` if no temp dir was created, and so you can be sure you can delete it if it's not `None`?

Sephiroth-Lin replied:
@srowen sorry, I don't very clear. You mean we can not use the executorId to distinguish the driver and executor?

srowen replied:
I mean that this is not exactly the same condition that decided whether to make a temp dir, although, it looks like it will end up being the same. Since the consequence of accidentally getting this wrong or due to a later change, I suggest making this much more intimately bound. For example: save a reference to the temp dir if it exists. That you know you can delete.

Sephiroth-Lin replied:
@srowen Than you. If we want to make this much more intimately bound, may be we can check the sparkFilesDir directly, or else, we need to add a parameter to SparkEnv class.

srowen replied:
I think you just need to store one more member, which is a reference to the tmp dir if it was created. See above. That's pretty much fool-proof and doesn't require any new params anywhere.

Sephiroth-Lin committed:
check sparkFilesDir before delete

change from < if (SparkContext.DRIVER_IDENTIFIER == executorId) > to < if (sparkFilesDir != ".") >, and add comment where sparkFilesDir is created.

srowen commented:
ok to test

SparkQA commented:
[Test build #27097 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27097/consoleFull) for   PR 4412 at commit [`b2018a5`](https://github.com/apache/spark/commit/b2018a5996ad10ef5938bab1dfd9d0dfbc178a56).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27097 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27097/consoleFull) for   PR 4412 at commit [`b2018a5`](https://github.com/apache/spark/commit/b2018a5996ad10ef5938bab1dfd9d0dfbc178a56).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27097/
Test PASSed.

Sephiroth-Lin committed:
add a member to store the reference of tmp dir

SparkQA commented:
[Test build #27186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27186/consoleFull) for   PR 4412 at commit [`1339c96`](https://github.com/apache/spark/commit/1339c969af36f94eb7c07e96ee6459373b7678f9).
 * This patch merges cleanly.

Sephiroth-Lin commented:
@srowen thank you, now I add a member to store the reference of the tmp dir if it was created, please help to check again.

SparkQA commented:
[Test build #27186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27186/consoleFull) for   PR 4412 at commit [`1339c96`](https://github.com/apache/spark/commit/1339c969af36f94eb7c07e96ee6459373b7678f9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27186/
Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/HttpFileServer.scala:
```diff
@@ -50,6 +50,15 @@ private[spark] class HttpFileServer(
+      Utils.deleteRecursively(baseDir)
+    } catch {
+      case e: Exception =>
+        logWarning("Exception while deleting Spark temp dir: " + baseDir.getAbsolutePath, e)
```
Nit: consider interpolation here

Sephiroth-Lin replied:
OK.

srowen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -76,6 +76,8 @@ class SparkEnv (
   // (e.g., HadoopFileRDD uses this to cache JobConfs and InputFormats).
   private[spark] val hadoopJobMetadata = new MapMaker().softValues().makeMap[String, Any]()
 
+  private var tmpFilesDir: Option[String] = None
```
Yeah this looks good. Maybe call it `driverTmpDirToDelete` or something to be extra clear?

srowen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -367,6 +385,13 @@ object SparkEnv extends Logging {
       conf)
+      
+    // Add a reference to tmp dir created by driver
+    if (isDriver) {
```
Might call out why this is linked to the `isDriver` logic above for future readers. @JoshRosen thoughts now?

Sephiroth-Lin committed:
rename the variable and update comment

Sephiroth-Lin committed:
using an interpolated string

SparkQA commented:
[Test build #27199 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27199/consoleFull) for   PR 4412 at commit [`b968e14`](https://github.com/apache/spark/commit/b968e148655f108b7501573bc4958369c7e17238).
 * This patch merges cleanly.

Sephiroth-Lin commented:
@srowen thank you, please help to check again.

SparkQA commented:
[Test build #27199 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27199/consoleFull) for   PR 4412 at commit [`b968e14`](https://github.com/apache/spark/commit/b968e148655f108b7501573bc4958369c7e17238).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27199/
Test PASSed.

Sephiroth-Lin committed:
using an interpolated string

SparkQA commented:
[Test build #27200 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27200/consoleFull) for   PR 4412 at commit [`fbbc785`](https://github.com/apache/spark/commit/fbbc7857816aae5ce0c5d900783485f938f28447).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27200 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27200/consoleFull) for   PR 4412 at commit [`fbbc785`](https://github.com/apache/spark/commit/fbbc7857816aae5ce0c5d900783485f938f28447).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27200/
Test PASSed.

srowen commented:
OK, that's looking good to me. I'll pause shortly to let others review it too.
